### ISSUE TITLE
[bug] string contains null byte

### ISSUE THREAD

**Issue by fnxpt** [sparklemotion/nokogiri#3385]
Im executing `cyclonedx-cocoapods` to generate a sbom for an iOS application, this uses nokogiri to generate the xml file. I notice that with version 1.17.2 it fails with the following error

```
string contains null byte
/.gem/ruby/3.3.0/gems/nokogiri-1.17.2-arm64-darwin/lib/nokogiri/xml/document.rb:307:in `new'
/.gem/ruby/3.3.0/gems/nokogiri-1.17.2-arm64-darwin/lib/nokogiri/xml/document.rb:307:in `create_cdata'
/.gem/ruby/3.3.0/gems/nokogiri-1.17.2-arm64-darwin/lib/nokogiri/xml/builder.rb:346:in `cdata'
/.gem/ruby/3.3.0/gems/cyclonedx-cocoapods-1.4.1/lib/cyclonedx/cocoapods/bom_builder.rb:130:in `block (2 levels) in add_to_bom'
```

this doesn't happen with version 1.13

**Expected behavior**

It should be able to handle null bytes

**Environment**
# Nokogiri (1.17.2)
    ---
    warnings: []
    nokogiri:
      version: 1.17.2
      cppflags:
      - "-I/Users/user/.gem/ruby/3.3.0/gems/nokogiri-1.17.2-arm64-darwin/ext/nokogiri"
      - "-I/Users/user/.gem/ruby/3.3.0/gems/nokogiri-1.17.2-arm64-darwin/ext/nokogiri/include"
      - "-I/Users/user/.gem/ruby/3.3.0/gems/nokogiri-1.17.2-arm64-darwin/ext/nokogiri/include/libxml2"
      ldflags: []
    ruby:
      version: 3.3.6
      platform: arm64-darwin23
      gem_platform: arm64-darwin-23
      description: ruby 3.3.6 (2024-11-05 revision 75015d4c1f) [arm64-darwin23]
      engine: ruby
    libxml:
      source: packaged
      precompiled: true
      patches:
      - 0001-Remove-script-macro-support.patch
      - 0002-Update-entities-to-remove-handling-of-ssi.patch
      - '0009-allow-wildcard-namespaces.patch'
      - 0010-update-config.guess-and-config.sub-for-libxml2.patch
      - 0011-rip-out-libxml2-s-libc_single_threaded-support.patch
      memory_management: ruby
      iconv_enabled: true
      compiled: 2.13.5
      loaded: 2.13.5
    libxslt:
      source: packaged
      precompiled: true
      patches:
      - 0001-update-config.guess-and-config.sub-for-libxslt.patch
      datetime_enabled: true
      compiled: 1.1.42
      loaded: 1.1.42
    other_libraries:
      zlib: 1.3.1
      libiconv: '1.17'
      libgumbo: 1.0.0-nokogiri

**Additional context**

CycloneDX/cyclonedx-cocoapods#85

**Comment by flavorjones** [comment]
Hi, thanks for opening this issue.

This behavior changed in commit a53df25f in which we ensure the argument to the `CData` constructor is a well-formed C string. We were forced to change this because previously, `CData` erroneously allowed non-string arguments (which allowed your program to run with earlier versions), and the upstream change in gnome/libxml2@9991fae4 caught that bug.

If it's helpful, this is likely a bug in your application, since an embedded null byte in a C string is a terminator:

```ruby
# with Nokogiri < v1.17.0
pp Nokogiri::XML::Document.new.create_cdata("first\0last")
# => #(CDATA "first")
```

meaning that you're not parsing the full input string even in older versions.

Here's some historical information if it's helpful:

- the XML standard's allowed character set does not include the NULL byte: https://www.w3.org/TR/REC-xml/#charsets
- starting in Nokogiri 1.6.8 we did work to ensure that input strings did not contain NULL bytes, but missed CData
- a previous issue about null-byte handling in 1.6.8: https://github.com/sparklemotion/nokogiri/issues/1429

TL;DR: libxml2 is a C library which means it never handled strings with null characters. We changed Nokogiri's behavior in v1.6.8 in most places but not CData. We were forced to fix CData in v1.17.0. This change is now raising an exception to let you know your data can't be parsed correctly.

I hope this answer is helpful.

**Comment by fnxpt** [comment]
Thanks for the feedback, any advise on how to include this values... Im discussing this, and in my opinion maybe the data should be encoded... for reference the string we are trying to add is description field from RXSwift -> 
```This is a Swift port of [ReactiveX.io](https://github.com/ReactiveX)

Like the original [Rx](https://github.com/Reactive-extensions/Rx.Net), its intention is to enable easy composition of asynchronous operations and event streams.

It tries to port as many concepts from the original Rx as possible, but some concepts were adapted for more pleasant and performant integration with iOS/macOS/Linux environment.

Probably the best analogy for those who have never heard of Rx would be:

```
git diff | grep bug | less          #  linux pipes - programs communicate by sending
				    #  sequences of bytes, words, lines, '\0' terminated strings...
```
would become if written in RxSwift
```
gitDiff().grep("bug").less          // sequences of swift objects
```

**Comment by flavorjones** [comment]
@fnxpt I've got no idea what your users expect to see in the final output when the input data is invalid like this.

What I can tell you is that they previously weren't getting anything after the NULL byte, so one thing you could do to preserve functionality is just chop off everything from the null byte onwards:

```ruby
original = "asdf\0qwer"
modified = original[...original.index("\0")] # => "asdf"
```

Maybe you could just always encode it using `String#dump`?

```
original = "asdf\0qwer"
original.dump # => "asdf\x00qwer"

Nokogiri::XML::Document.new.create_cdata(x.dump)
# => #(CDATA "\"asdf\\x00qwer\"")

Nokogiri::XML::Document.new.create_cdata(x.dump).to_xml
# => "<![CDATA[\"asdf\\x00qwer\"]]>"
```

Like I said, because I don't know what your users want in this case, I can only guess.

**Comment by flavorjones** [comment]
Or, like, just gsub it?

```ruby
>> x = "asdf\0qwer"
>> x.gsub("\0", "\\x00")
=> "asdf\\x00qwer"
```

**Comment by flavorjones** [comment]
Ah, wait, ok, are you saying that this text

```
#  sequences of bytes, words, lines, '\0' terminated strings...
```

is being fed into Nokogiri with a null byte in it? If so, that's a bug in how you're processing this input.

Please be careful about whether you're using data that is being string-interpolated by Ruby. Here's the difference:

```ruby
# interpolated, the characters "\" and "0" are interpreted as a null byte
x = "lines, '\0' terminated"
# => "lines, '\u0000' terminated"

# not interpolated, "\" and "0" remain untouched
x = %q"lines, '\0' terminated"
# => "lines, '\\0' terminated"
```

TL;DR there's no null byte in the data you posted above, and if there is it's an issue in how you're processing it before it gets to Nokogiri.
